### PR TITLE
Graceful failed response handling

### LINKED ISSUE #4: JSON channel should handle a failed response gracefully

### PR DESCRIPTION
**Pull request by rok-clarifai**
This PR makes the failed or mixed response in a JSON channel to be handled in the same way as in the gRPC channel.

It also improves tests' structure.

Closes #4 .

### CODE REVIEW

File: tests/helpers.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      raise Exception(
+        response.status.description + " " + response.status.details
+      )
+  # At this point, the model has successfully finished evaluation.
+
+
+def _raise_on_failure(response):
+  if response.status.code != status_code_pb2.SUCCESS:
+    print("Received failure response:")
+    print(response)
```
Comment (JeckyTian) [inline tests/helpers.py:87]: This is for debugging? 
I'm thinking if the info in exception is enough for engineers to debug api client test failures. If we want to encode all "response" into exception?

File: tests/helpers.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        service_pb2.GetInputRequest(input_id=input_id),
+        metadata=metadata
+      )
+      _raise_on_failure(get_input_response)
+      if get_input_response.input.status.code == status_code_pb2.INPUT_DOWNLOAD_SUCCESS:
+        break
+      elif get_input_response.input.status.code in (status_code_pb2.INPUT_DOWNLOAD_PENDING,
+                                                    status_code_pb2.INPUT_DOWNLOAD_IN_PROGRESS):
+        time.sleep(1)
+        continue
```
Comment (JeckyTian) [inline tests/helpers.py:36]: I guess this `continue` is not needed.

File: tests/helpers.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      else:
+        raise Exception(
+          get_input_response.status.description + " " + get_input_response.status.details
```
Comment (JeckyTian) [inline tests/helpers.py:39]: can we also add `status.code`?

File: tests/helpers.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      service_pb2.GetModelVersionRequest(model_id=model_id, version_id=model_version_id),
+      metadata=metadata
+    )
+    _raise_on_failure(response)
+    if response.model_version.status.code == status_code_pb2.MODEL_TRAINED:
+      break
+    elif response.model_version.status.code in (status_code_pb2.MODEL_QUEUED_FOR_TRAINING,
+                                                status_code_pb2.MODEL_TRAINING):
+      time.sleep(1)
+      continue
```
Comment (JeckyTian) [inline tests/helpers.py:56]: no need?

File: tests/helpers.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    else:
+      raise Exception(
+        response.status.description + " " + response.status.details
```
Comment (JeckyTian) [inline tests/helpers.py:59]: may also want `code` here?  to make debugging easy

File: tests/helpers.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      service_pb2.GetModelVersionMetricsRequest(model_id=model_id, version_id=model_version_id),
+      metadata=metadata
+    )
+    _raise_on_failure(response)
+    if response.model_version.metrics.status.code == status_code_pb2.MODEL_EVALUATED:
+      break
+    elif response.model_version.metrics.status.code in (status_code_pb2.MODEL_QUEUED_FOR_EVALUATION,
+                                                        status_code_pb2.MODEL_EVALUATING):
+      time.sleep(1)
+      continue
```
Comment (JeckyTian) [inline tests/helpers.py:76]: Hi~

File: tests/helpers.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    else:
+      raise Exception(
+        response.status.description + " " + response.status.details
```
Comment (JeckyTian) [inline tests/helpers.py:79]: Hi again~

File: tests/test_integration.py [hunk 8]
```diff
@@ -105,10 +123,29 @@ def _assert_post_patch_delete_input(channel):
     assert status_code_pb2.SUCCESS == delete_response.status.code
 
 
-def _assert_multiple_requests(channel):
+@both_channels
+def test_list_models_with_pagination(channel):
+  stub = service_pb2_grpc.V2Stub(channel)
+
+  response = stub.ListModels(service_pb2.ListModelsRequest(per_page=2), metadata=metadata)
+  if response.status.code != status_code_pb2.SUCCESS:
```
Comment (JeckyTian) [inline tests/test_integration.py:131]: this is also `_raise_on_failure`

File: tests/test_integration.py [hunk 9]
```diff
@@ -105,10 +123,29 @@ def _assert_post_patch_delete_input(channel):
...
+    raise Exception(response.status.description + " " + response.status.details)
+  assert len(response.models) == 2
+
+  # We shouldn 't have 1000*500 number of models, so the result should be empty.
+  response = stub.ListModels(
+    service_pb2.ListModelsRequest(page=1000, per_page=500),
+    metadata=metadata
+  )
+  if response.status.code != status_code_pb2.SUCCESS:
```
Comment (JeckyTian) [inline tests/test_integration.py:140]: `_raise_on_failure`?

File: tests/helpers.py (hunk 1 above)
Comment (rok-clarifai) [reply tests/helpers.py]: Yes, I added the whole response object here.

**Comment by rok-clarifai** [comment]
@JeckyTian Thanks for your comments, I addressed them. It's a public library, it will be used internally as well.

**Pull request merged**